include/deal.II/lac/read_write_vector.h

```diff
@@ -0,0 +1,730 @@
+     */
+    const IndexSet &get_stored_elements () const;
+
+    /**
+     * Make the @p ReadWriteVector class a bit like the <tt>vector<></tt> class of
+     * the C++ standard library by returning iterators to the start and end
+     * of the <i>locally stored</i> elements of this vector.
+     */
+    iterator begin ();
+
+    /**
+     * Returns constant iterator to the start of the locally stored elements
+     * of the vector.
+     */
+    const_iterator begin () const;
+
+    /**
+     * Returns an iterator pointing to the element past the end of the array
+     * of locally stored entries.
+     */
+    iterator end ();
+
+    /**
+     * Returns a constant iterator pointing to the element past the end of
+     * the array of the locally stored entries.
+     */
+    const_iterator end () const;
+    //@}
+
+
+    /**
+     * @name 2: Data-Access
+     */
+    //@{
+
+    /**
+     * Read access to the data in the position corresponding to @p
+     * global_index.
+     */
+    Number operator () (const size_type global_index) const;
+
+    /**
+     * Read and write access to the data in the position corresponding to @p
+     * global_index.
+     */
+    Number &operator () (const size_type global_index);
+
+    /**
+     * Read access to the data in the position corresponding to @p
+     * global_index.
+     *
+     * This function does the same thing as operator().
+     */
+    Number operator [] (const size_type global_index) const;
+
+    /**
+     * Read and write access to the data in the position corresponding to @p
+     * global_index.
+     *
+     * This function does the same thing as operator().
+     */
+    Number &operator [] (const size_type global_index);
+
+    /**
+     * Instead of getting individual elements of a vector, this function
+     * allows to get a whole set of elements at once. The indices of the
+     * elements to be read are stated in the first argument, the
+     * corresponding values are returned in the second.
+     */
+    template <typename Number2>
+    void extract_subvector_to (const std::vector<size_type> &indices,
+                               std::vector<Number2> &values) const;
+
+    /**
+     * Just as the above, but with pointers. Useful in minimizing copying of
+     * data around.
+     */
+    template <typename ForwardIterator, typename OutputIterator>
+    void extract_subvector_to (ForwardIterator          indices_begin,
+                               const ForwardIterator    indices_end,
+                               OutputIterator           values_begin) const;
+
+    /**
+     * Read access to the data field specified by @p local_index. The largest
+     * range of the IndexSet is the first one.
+     *
+     * Performance: Direct array access (fast).
+     */
+    Number local_element (const size_type local_index) const;
+
+    /**
+     * Read and write access to the data field specified by @p local_index. The
+     * largest range of the IndexSet is the first one.
```

Comment:
You are here (and in the previous function) leaking information about the internal implementation of the class. This may be better suited for the documentation of the class as a whole. Maybe replace the second sentence above by this:
```
  When you access elements in the order in which they are stored, it is necessary that you know in which order this class stores those its elements. In other words, you need to know the map between the global indices of the elements this class stores, and the local indices into the contiguous array of these global elements. For this, see the general documentation of this class.
```
Then, in the class doc, say something like this:
```
<h3>Storing elements</h3>

This class is intended to represent vectors in ${\mathbb R}^N$ for which it stores all or a subset of elements. The latter case is important in parallel computations, where $N$ may be so large that no processor can actually store all elements of a solution vector, but where this is also not necessary: one typically only has to store the values of degrees of freedom that live on cells that are locally owned, plus potentially those degrees of freedom that live on ghost cells.

Most of the time, one will simply read from or write into a vector of the current class using the global numbers of these degrees of freedom. This is done using operator() or operator[]. In such cases, it is clear that one can only access elements of the vector that the current object indeed stores.

However, it is also possible to access elements in the order in which they are stored by the current object. In other words, one is not interested in accessing elements with their <i>global</i> indices, but instead using an enumeration that only takes into account the elements that are actually stored. This is facilitated by the local_element() function. To this end, it is necessary to know <i>in which order</i> the current class stores its element. This order is defined as follows:
- First all of the elements of the largest consecutive range of indices that this vector stores.
- Then the elements of all other consecutive ranges, in ascending order of the first index of each range.
```